server/portal/apps/projects/views.py

```diff
@@ -86,8 +86,14 @@ def get(self, request):
             search = search.extra(from_=int(offset), size=int(limit))
 
             res = search.execute()
-            hits = [hit.to_dict() for hit in res]
-            listing = hits
+            hits = list(map(lambda hit: hit.id, res))
+            listing = []
+            # Filter search results to projects specific to user
+            if hits:
+                client = request.user.tapis_oauth.client
+                listing = list_projects(client)
+                filtered_list = filter(lambda prj: prj['id'] in hits, listing)
```

Comment:
[not tested or reviewed] but would filtering the results break the usage of the offset/limit params?   like a user would get filtered results that were less than the `limit` so would assume that there are no more search results when there could be more search results.

Reply:
@nathanfranklin  
Yes, that is a good point. It only shows within the offset. list_projects by default gets results by offset - 50. And so, it will have results less than search results. 
The search index has all user results.
Using settings.PORTAL_PROJECTS_SYSTEM_PREFIX (mentioned in comment above), will restrict system and NOT user. So, it will still return result for all users.

So, any search will have to check if a workspace is available to a user. Access controls are handled by tapis, for example: is a workspace owned, or shared, and access control logic is handled by tapis. 
So, to provide accurate results - 
1) we use tapis results - retrieve all workspaces without a filter for a user. 
   Pros: This will work.
   Cons: Defeats the purpose of offset and search. Bad perf.
2) Replicate the access control logic for search in portal. Use the search results and filter it locally for the current user. I haven't checked how sharing is organized.
   Pros: better perf
   Cons: Can sharing be checked and duplication of access control logic.

Also, please let me know if I'm wrong about system prefix based search. That only filters a system and not user. It is tough to test this locally, with only one user in my local dev env.

Reply:
Regarding data files search mentioned above: there is a [listing request](https://github.com/TACC/Core-Portal/blob/b51c41f47412a9b21bb93fb30eb030086bbdfcd4/server/portal/libs/agave/operations.py#L115) sent to tapis to ensure user has access before proceeding to search. So, it does search and tapis listing.

Reply:
I tested with filtering on system prefix. The good part about this is it does not crash the UI, just as suspected the previous result was brining in results from older system, which had different index and hence the crash.

And as expected, it is bringing in results from other users in the system. Even though I have no access, i can see others projects listed. I have 3 projects that match `test` keyword, but it brings in all other projects.

![Screenshot 2023-11-02 at 6 55 43 PM](https://github.com/TACC/Core-Portal/assets/2568355/22706aca-47ab-4a31-944a-f7d1d8e83a1b)

One possible solution to this: Store user metadata in the index, in addition to owner, store list of users with access to this. Essentially, something like this: owner: {}, users: [{}]. 
And I do not see tapis api listing all other users with access. So, everytime a user accesses this, the index needs to be updated with user added to the list.
The old index for project stored - PIs and CO-PIs.

The only downside of this (or for every elastic search index): the data consistency between index and db. If tapis is updated - a project is unshared, but not via portal, then index could go out of sync. I do not know if this is an important scenario, but something for us to know, in case a portal has strictness on data access controls.

Note: I'll bring back the previous code which cross checks tapis with search results, and keeping the system check.

Reply:
Since we are only filtering the index based on the user query and system prefix I can see the issue that Chandra encountered happening where results from other users are returned. I don't think it would be feasible to try to replicate and manage the system access control in the index and would rather have tapis remain the source of truth.  So I think Chandra's current solution with getting a list of the users projects should work fine. 

With regards to the offset and limit stuff, I don't think we are doing any kind of pagination on the project listing page so the offset and limit are not really needed. For reference there is no offset or limit parameter sent in [fetchProjectListing](https://github.com/TACC/Core-Portal/blob/main/client/src/redux/sagas/projects.sagas.js#L5) and on the project listing page there is nothing configured for [infinite scrolling](https://github.com/TACC/Core-Portal/blob/main/client/src/components/DataFiles/DataFilesProjectsList/DataFilesProjectsList.jsx#L22). The default tapis limit when getting systems is 100 so I guess that's the only one used. [This](https://github.com/TACC/Core-Portal/blob/main/server/portal/apps/projects/views.py#L86) line might not even be needed at all.

Reply:
excellent.  so drop offset/limit from endpoint 👍 

> The default tapis limit when getting systems is 100

So we need to check if `list_projects` returns a subset or really all of the projects.

Reply:
`list_projects` does a tapis `getSystems` call which by default returns 100 projects which should be enough I think? Also according to the [docs](https://tapis-project.github.io/live-docs/?service=Systems#tag/Systems/operation/getSystems) we can set the limit to -1 to get all the projects which might work better for us

Reply:
> according to the [docs](https://tapis-project.github.io/live-docs/?service=Systems#tag/Systems/operation/getSystems) we can set the limit to -1 to get all the projects which might work better for us

perfect. great that they have that built-in with the -1.  yeah, we'll definitely exceed 100 for a user at some point.

Reply:
What is the heuristic used in the past to decide whether offset/pagination is needed on a list? Should we check perf or suggestions from tapis team to not overload their services with large data requests? 
Regarding UI and response payload, the information returned for project is not huge like jobs - we have name, id, owner, and small metadata size. 
Each job has 90+ attributes vs project has < 10 attributes (250 B per project in payload). If that is a good enough criteria to justify using -1, then I can go ahead and use it. The memory size for 100 would be approx. 30 kB.

Reply:
Used -1 as [limit](https://tapis-project.github.io/live-docs/?service=Systems#tag/Systems/operation/getSystems).